drivers/secure_os/ws_api.cr

```diff
@@ -0,0 +1,204 @@
+require "placeos-driver"
+require "./ws_api_models"
+
+# docs: https://drive.google.com/file/d/1moo9NnFWukSf6fegxaZnSP5ShqSr_A03/view?usp=sharing
+
+class SecureOS::WsApi < PlaceOS::Driver
+  generic_name :SecureOS
+  descriptive_name "SecureOS WebSocket API"
+
+  uri_base "ws://secureos.server:8888/"
+  default_settings({
+    shared_host:   true,
+    rest_api_host: "http://172.16.1.120:8888",
+    basic_auth:    {
+      username: "srvc_acct",
+      password: "password!",
+    },
+    camera_types:  [] of String,
+    camera_states: [StateType::Attached, StateType::Armed, StateType::Alarmed],
+    camera_events: [] of String,
+  })
+
+  @rest_api_host : String = ""
+  @camera_list : Array(Camera) = [] of Camera
+  @camera_types : Array(String) = [] of String
+  @camera_states : Array(StateType) = [] of StateType
+  @camera_events : Array(String) = [] of String
+  @watchlist_list : Array(Watchlist) = [] of Watchlist
+
+  getter! basic_auth : NamedTuple(username: String, password: String)
+
+  def on_load
+    on_update
+  end
+
+  def on_update
+    @rest_api_host = setting String, :rest_api_host
+    @basic_auth = setting NamedTuple(username: String, password: String), :basic_auth
+    @camera_types = setting Array(String), :camera_types
+    @camera_states = setting Array(StateType), :camera_states
+    @camera_events = setting Array(String), :camera_events
+  end
+
+  def connected
+    host = setting?(Bool, :shared_host) ? config.uri.not_nil! : @rest_api_host
+    client = HTTP::Client.new URI.parse(host)
+    client.basic_auth **basic_auth
+    response = client.get "#{@rest_api_host}/api/v1/ws_auth"
```

Comment:
this code is repeated a few times, work creating a function so it's not repeated

```crystal
protected def http_client
  host = setting?(Bool, :shared_host) ? config.uri.not_nil! : @rest_api_host
  client = HTTP::Client.new URI.parse(host)
  client.basic_auth **basic_auth
  client
end
```

Then can do `response = http_client.get "#{@rest_api_host}/api/v1/ws_auth"`

Reply:
Yeah, that one kinda snuck up on me with more and more get/post requests slowly getting added.